### DISCUSSION TITLE
关于 stm 模块的疑问

### DISCUSSION THREAD

**Discussion by Gowa2017** [cloudwu/skynet#1474] [General]
在 [WIKI-STM](https://github.com/cloudwu/skynet/wiki/ShareData#stm) 中说到：

>stm.copy(stmobj) 可以从一个共享对象中生成一份读拷贝。它返回一个 stmcopy ，是一个 lightuserdata 。通常一定要把这个 lightuserdata 传到需要读取这份数据的服务。随意丢弃这个指针会导致内存泄露。注：一个拷贝只能供一个读取者使用，你不可以把这个指针传递给多个服务。如果你有多个读取者，为每个人调用 copy 生成一份读拷贝。

我看了一下代码，

```c
static int
lcopy(lua_State *L) {
	struct boxstm * box = lua_touserdata(L, 1);
	stm_grab(box->obj);
	lua_pushlightuserdata(L, box->obj);
	return 1;
}
```

只是简单的将指针传递给了 Lua，为什么会说：

**通常一定要把这个 lightuserdata 传到需要读取这份数据的服务。随意丢弃这个指针会导致内存泄露。**

**Comment by cloudwu** [comment]
stm 不维护了，不再建议使用。

**Reply by Gowa2017** [reply]
> stm 不维护了，不再建议使用。

那么云大，现在还有什么方式来进行在服务间数据共享么，我是说会有变更的数据。

当前， sharetable 是您在 wik 上推荐的方式，在您的博客上我也看到，那说的是你最终的尝试。

是否使用 sharetable 即可。

**Comment by cloudwu** [comment]
如果找不到非此不可的理由，不要用数据内存共享的方式同步数据。

发消息。

**Reply by Gowa2017** [reply]
> 如果找不到非此不可的理由，不要用数据内存共享的方式同步数据。
> 
> 发消息。

感谢。
